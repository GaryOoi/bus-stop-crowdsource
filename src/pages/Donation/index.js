import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";
import { Formik } from "formik";
import * as yup from "yup";
import {
  Container,
  Row,
  Col,
  Card,
  Form,
  Spinner,
  Modal,
} from "react-bootstrap";

import {
  addBusstopCurrentAmountRequest,
  addBusstopCurrentAmountSuccess,
  addBusstopCurrentAmountFail,
} from "actions/busstopActions";
import {
  getDonationsRequest,
  getDonationsSuccess,
  getDonationsFail,
  addDonationRequest,
  addDonationSuccess,
  addDonationFail,
} from "actions/donationActions";
import {
  getDonations,
  getDonationsLoading,
  getDonationsError,
} from "selectors/donationSelectors";
import { getSpecificBusstop } from "selectors/busstopSelectors";

import { MONTHS, YEARS } from "globals/constant";
import MessageModal from "components/MessageModal";
import UserDonation from "components/UserDonation";
import { getData, addDonation } from "utils/request";
import { genUUIDv4 } from "utils/utils";
import {
  Section,
  SectionHeader,
  DonationBar,
  DonationDescription,
  DonationFundedAmount,
  DonationIndicator,
  StyledProgressBar,
  HistoryWrapper,
  DonationHeader,
  DonationButton,
  SpinnerWrapper,
} from "./styles";

const initialState = {
  name: "",
  email: "",
  amount: "",
};

const bankInfoInitialState = {
  cardNumber: "",
  expirationMonth: "",
  expirationYear: "",
  cvv: "",
};

const schema = yup.object({
  name: yup.string().required("*Required"),
  email: yup.string().email("Invalid email"),
  amount: yup
    .string()
    .matches(/^\d+(?:\.\d{0,2})$/, "Invalid amount (eg: 12.00)")
    .required("*Required"),
});

const bankInfoSchema = yup.object({
  cardNumber: yup
    .string()
    .required("*Required")
    .matches(/^[0-9]+$/, "Must be only digits")
    .length(16, "Invalid Card Number")
    .max(16),
  expirationMonth: yup
    .string()
    .required("*Required")
    .matches(/^[0-9]+$/, "Must be only digits")
    .length(2, "Invalid Month"),
  expirationYear: yup
    .string()
    .required("*Required")
    .matches(/^[0-9]+$/, "Must be only digits")
    .length(4, "Invalid Year"),
  cvv: yup
    .string()
    .required("*Required")
    .matches(/^[0-9]+$/, "Must be only digits")
    .length(3, "Invalid CVV"),
});

function Donation({ match }) {
  const busStop = useSelector(getSpecificBusstop(match.params.busstopid));
  const percentage = parseFloat(
    (busStop.currentAmount / busStop.totalAmount) * 100
  ).toFixed(2);
  const leftAmount = busStop.totalAmount - busStop.currentAmount;
  const dispatch = useDispatch();
  const donations = useSelector(getDonations(match.params.busstopid));
  const donationsLoading = useSelector(getDonationsLoading);
  const donationsError = useSelector(getDonationsError);
  const [modelData, setModelData] = useState({
    visibility: false,
    message: "",
  });

  const [dataToSubmit, setDataToSubmit] = useState({
    isBankInfoModalVisible: false,
    message: "",
    values: {},
    actions: {},
  });

  const handlePressModal = () => {
    setModelData({
      ...modelData,
      visibility: !modelData.visibility,
    });
  };

  const handleBankInfoModalCancel = () => {
    dataToSubmit.actions.setSubmitting(false);
  };

  const handleSubmit = (values, actions) => {
    setDataToSubmit({
      values,
      actions,
    });
  };

  const handleBankInfoModalSubmit = async (values) => {
    try {
      dispatch(addDonationRequest());
      dispatch(addBusstopCurrentAmountRequest());
      // Process bank card
      await addDonation({ ...values, ...dataToSubmit.values });
      // Add donation
      const editedValues = {
        id: genUUIDv4(),
        name: dataToSubmit.values.name,
        email: dataToSubmit.values.email,
        amount: dataToSubmit.values.amount,
        busstopId: Number(match.params.busstopid),
        createdAt: new Date(),
      };
      const total =
        Number(dataToSubmit.values.amount) + Number(busStop.currentAmount);
      dispatch(addDonationSuccess(editedValues));
      dispatch(addBusstopCurrentAmountSuccess(match.params.busstopid, total));
    } catch (error) {
      dispatch(addDonationFail());
      dispatch(addBusstopCurrentAmountFail());
      setModelData({
        visibility: true,
        message: error,
      });
    }
  };

  useEffect(() => {
    const fetchAPI = async () => {
      try {
        dispatch(getDonationsRequest());
        const result = await getData("donations");

        dispatch(getDonationsSuccess(result));
      } catch (err) {
        dispatch(getDonationsFail(err));
        setModelData({
          visibility: true,
          message: err,
        });
      }
    };
    if (donations.length < 1) {
      fetchAPI();
    }
  }, [donations.length, dispatch]);

  return (
    <>
      <Section>
        {donationsLoading || donationsError !== "" ? (
          <SpinnerWrapper>
            <Spinner animation="border" role="status">
              <span className="sr-only">Loading...</span>
            </Spinner>
          </SpinnerWrapper>
        ) : (
          <Container>
            <SectionHeader className="mb-5">{busStop.name}</SectionHeader>
            <Row>
              <Col lg={7}>
                <div className="mb-4">
                  <DonationBar className="mb-3" />
                  <DonationFundedAmount>
                    ${busStop.currentAmount}
                  </DonationFundedAmount>
                  <DonationIndicator className="mb-2">
                    pledged of ${busStop.totalAmount} goal
                  </DonationIndicator>
                  <StyledProgressBar
                    now={percentage}
                    label={`${percentage}%`}
                  />
                </div>
                <Row>
                  <Col>
                    <DonationDescription>
                      ${leftAmount > 0 ? leftAmount : 0}
                    </DonationDescription>
                    <DonationIndicator>More to go</DonationIndicator>
                  </Col>
                  <Col>
                    <div className="mb-5">
                      <DonationDescription>
                        {donations.length}
                      </DonationDescription>
                      <DonationIndicator>Backers</DonationIndicator>
                    </div>
                  </Col>
                </Row>

                <h4 className="mb-4">History</h4>
                <HistoryWrapper>
                  {donations.map((donation) => {
                    return (
                      <UserDonation key={donation.id} donation={donation} />
                    );
                  })}
                </HistoryWrapper>
              </Col>
              <Col lg={5}>
                <Card>
                  <Card.Body>
                    <DonationHeader className="font-weight-bold mb-4">
                      Fund this bus stop
                    </DonationHeader>
                    <Formik
                      validationSchema={schema}
                      onSubmit={handleSubmit}
                      initialValues={initialState}
                    >
                      {({
                        handleSubmit,
                        handleChange,
                        values,
                        errors,
                        touched,
                        isSubmitting,
                      }) => (
                        <Form noValidate onSubmit={handleSubmit}>
                          <Form.Group>
                            <Form.Label>Name*</Form.Label>
                            <Form.Control
                              type="text"
                              name="name"
                              value={values.name}
                              onChange={handleChange}
                              isInvalid={touched.name && !!errors.name}
                              placeholder="Enter your name"
                            />
                            <Form.Control.Feedback type="invalid">
                              {errors.name}
                            </Form.Control.Feedback>
                          </Form.Group>
                          <Form.Group controlId="formBasicEmail">
                            <Form.Label>Email address</Form.Label>
                            <Form.Control
                              type="email"
                              name="email"
                              value={values.email}
                              onChange={handleChange}
                              isInvalid={touched.email && !!errors.email}
                              placeholder="Enter your email"
                            />
                            <Form.Text className="text-muted">
                              We'll never share your email with anyone else.
                            </Form.Text>
                            <Form.Control.Feedback type="invalid">
                              {errors.email}
                            </Form.Control.Feedback>
                          </Form.Group>

                          <Form.Group
                            className="mb-5"
                            controlId="formBasicEmail"
                          >
                            <Form.Label>Fund Amount*</Form.Label>
                            <Form.Control
                              type="string"
                              name="amount"
                              value={values.amount}
                              onChange={handleChange}
                              isInvalid={touched.amount && !!errors.amount}
                              placeholder="Enter amount"
                            />
                            <Form.Control.Feedback type="invalid">
                              {errors.amount}
                            </Form.Control.Feedback>
                          </Form.Group>

                          <DonationButton type="submit" size="lg" block>
                            Fund Now
                          </DonationButton>
                          <Modal
                            show={isSubmitting}
                            onHide={handleBankInfoModalCancel}
                          >
                            <Modal.Header closeButton>
                              <Modal.Title>
                                Bank Account Information
                              </Modal.Title>
                            </Modal.Header>
                            <Formik
                              validationSchema={bankInfoSchema}
                              onSubmit={handleBankInfoModalSubmit}
                              initialValues={bankInfoInitialState}
                            >
                              {({
                                handleSubmit,
                                handleChange,
                                values,
                                errors,
                                touched,
                              }) => (
                                <Form noValidate onSubmit={handleSubmit}>
                                  <Modal.Body>
                                    <Form.Group>
                                      <Form.Label>Card Number*</Form.Label>
                                      <Form.Control
                                        type="text"
                                        name="cardNumber"
                                        value={values.cardNumber}
                                        onChange={handleChange}
                                        isInvalid={
                                          touched.cardNumber &&
                                          !!errors.cardNumber
                                        }
                                        placeholder="1234 4568 9101 1121"
                                      />
                                      <Form.Control.Feedback type="invalid">
                                        {errors.cardNumber}
                                      </Form.Control.Feedback>
                                    </Form.Group>

                                    <Form.Row>
                                      <Form.Group
                                        as={Col}
                                        controlId="formGridCity"
                                      >
                                        <Form.Label>
                                          Expiration Date*
                                        </Form.Label>
                                        <Form.Control
                                          as="select"
                                          name="expirationMonth"
                                          value={values.expirationMonth}
                                          onChange={handleChange}
                                          isInvalid={
                                            touched.expirationMonth &&
                                            !!errors.expirationMonth
                                          }
                                        >
                                          <option value="">Month</option>
                                          {MONTHS.map((month) => {
                                            return (
                                              <option key={month} value={month}>
                                                {month}
                                              </option>
                                            );
                                          })}
                                        </Form.Control>
                                        <Form.Control.Feedback type="invalid">
                                          {errors.expirationMonth}
                                        </Form.Control.Feedback>
                                      </Form.Group>

                                      <Form.Group
                                        as={Col}
                                        controlId="formGridState"
                                      >
                                        <Form.Label>&nbsp; </Form.Label>
                                        <Form.Control
                                          as="select"
                                          name="expirationYear"
                                          value={values.expirationYear}
                                          onChange={handleChange}
                                          isInvalid={
                                            touched.expirationYear &&
                                            !!errors.expirationYear
                                          }
                                        >
                                          <option value="">Year</option>
                                          {YEARS.map((year) => {
                                            return (
                                              <option key={year} value={year}>
                                                {year}
                                              </option>
                                            );
                                          })}
                                        </Form.Control>
                                        <Form.Control.Feedback type="invalid">
                                          {errors.expirationYear}
                                        </Form.Control.Feedback>
                                      </Form.Group>

                                      <Form.Group
                                        className="mb-5"
                                        as={Col}
                                        controlId="formGridZip"
                                      >
                                        <Form.Label>CVV*</Form.Label>
                                        <Form.Control
                                          name="cvv"
                                          value={values.cvv}
                                          onChange={handleChange}
                                          isInvalid={
                                            touched.cvv && !!errors.cvv
                                          }
                                          placeholder="123"
                                        />
                                        <Form.Control.Feedback type="invalid">
                                          {errors.cvv}
                                        </Form.Control.Feedback>
                                      </Form.Group>
                                    </Form.Row>
                                  </Modal.Body>
                                  <Modal.Footer>
                                    <DonationButton type="submit">
                                      Submit
                                    </DonationButton>
                                  </Modal.Footer>
                                </Form>
                              )}
                            </Formik>
                          </Modal>
                        </Form>
                      )}
                    </Formik>
                  </Card.Body>
                </Card>
              </Col>
            </Row>
          </Container>
        )}
      </Section>
      <MessageModal modelData={modelData} onClick={handlePressModal} />
    </>
  );
}

Donation.propTypes = {
  match: PropTypes.object.isRequired,
};

export default Donation;
