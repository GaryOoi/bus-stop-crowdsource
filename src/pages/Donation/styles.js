import styled from "styled-components";
import { ProgressBar, Image, Button } from "react-bootstrap";

export const Section = styled.div`
  padding: 3em 0;
  position: relative;
  @media (max-width: 767.98px) {
    padding-top: 4em;
  }
`;

export const SectionHeader = styled.h2`
  font-size: 30px;
  font-weight: 800;
`;

export const DonationBar = styled.div`
  height: 8px;
  width: 10%;
  background-color: #009e74;
`;

export const DonationDescription = styled.h3`
  font-size: 28px;
`;

export const DonationFundedAmount = styled(DonationDescription)`
  color: #009e74;
`;

export const DonationIndicator = styled.div`
  font-size: 14px;
`;

export const StyledProgressBar = styled(ProgressBar)`
  .progress-bar {
    background-color: unset;
    background: linear-gradient(210deg, #5de455 0%, #14c0aa 100%) !important;
  }
`;

export const UserWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
`;

export const StyledImg = styled(Image)`
  width: 40px;
`;

export const UserDescription = styled.div`
  color: #757575;
`;

export const HistoryWrapper = styled.div`
  overflow-y: scroll;
  height: 400px;
`;

export const DonationHeader = styled.h4`
  color: #009e74;
`;

export const DonationButton = styled(Button)`
  background-color: #009e74;
  color: #fff;
  font-weight: 700;
  border: 0;

  &:hover,
  &:active {
    background-color: #037362;
    color: #fff;
  }
`;

export const SpinnerWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;
