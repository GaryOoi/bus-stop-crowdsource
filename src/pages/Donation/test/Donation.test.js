import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

import Donation from "../index";

test("renders learn react link", () => {
  const mockStore = configureStore([]);
  const store = mockStore({
    busstop: {
      busstops: [
        {
          id: 1,
          name: "xyz",
          currentAmount: 6,
          totalAmount: 700,
        },
      ],
    },
    donation: {
      isLoading: false,
      error: "",
      donations: [
        {
          id: 1,
          name: "ali",
          email: "ali@yahoo.com",
          amount: 50,
          busstopId: 1,
          createdAt: "2020-07-12T14:55:31.173+00:00",
        },
      ],
    },
  });

  const match = {
    params: { busstopid: "1" },
  };

  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Donation match={match} />
      </MemoryRouter>
    </Provider>
  );

  expect(container.firstChild).toMatchSnapshot();
});
