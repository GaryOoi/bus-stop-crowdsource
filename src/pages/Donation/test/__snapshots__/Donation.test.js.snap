// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  class="sc-AxhCb gCfWfU"
>
  <div
    class="container"
  >
    <h2
      class="sc-AxhUy fzfMcb mb-5"
    >
      xyz
    </h2>
    <div
      class="row"
    >
      <div
        class="col-lg-7"
      >
        <div
          class="mb-4"
        >
          <div
            class="sc-AxgMl fVmbZT mb-3"
          />
          <h3
            class="sc-AxheI sc-Axmtr kIipBh"
          >
            $
            6
          </h3>
          <div
            class="sc-AxmLO QBcTe mb-2"
          >
            pledged of $
            700
             goal
          </div>
          <div
            class="sc-fzozJi cJMJbw progress"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0.86"
              class="progress-bar"
              role="progressbar"
              style="width: 0.86%;"
            >
              0.86%
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <h3
              class="sc-AxheI icydrP"
            >
              $
              694
            </h3>
            <div
              class="sc-AxmLO QBcTe"
            >
              More to go
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="mb-5"
            >
              <h3
                class="sc-AxheI icydrP"
              >
                1
              </h3>
              <div
                class="sc-AxmLO QBcTe"
              >
                Backers
              </div>
            </div>
          </div>
        </div>
        <h4
          class="mb-4"
        >
          History
        </h4>
        <div
          class="sc-fznyAO dxLNYP"
        >
          <div
            class="sc-AxjAm fUWBhh mb-4"
          >
            <img
              class="sc-AxirZ sYfwZ mr-3 rounded"
              src="user.svg"
            />
            <div
              class="sc-AxiKw eBvrQH"
            >
              <strong>
                ali
              </strong>
               donated $
              50
               on 
              July 12, 2020
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-lg-5"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <h4
              class="sc-fznKkj KRuEy font-weight-bold mb-4"
            >
              Fund this bus stop
            </h4>
            <form
              class=""
              novalidate=""
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                >
                  Name*
                </label>
                <input
                  class="form-control"
                  name="name"
                  placeholder="Enter your name"
                  type="text"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="formBasicEmail"
                >
                  Email address
                </label>
                <input
                  class="form-control"
                  id="formBasicEmail"
                  name="email"
                  placeholder="Enter your email"
                  type="email"
                  value=""
                />
                <small
                  class="text-muted form-text"
                >
                  We'll never share your email with anyone else.
                </small>
                <div
                  class="invalid-feedback"
                />
              </div>
              <div
                class="mb-5 form-group"
              >
                <label
                  class="form-label"
                  for="formBasicEmail"
                >
                  Fund Amount*
                </label>
                <input
                  class="form-control"
                  id="formBasicEmail"
                  name="amount"
                  placeholder="Enter amount"
                  type="string"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
              <button
                class="sc-fznZeY jLpRNd btn btn-primary btn-block btn-lg"
                type="submit"
              >
                Fund Now
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
